# Аналитический отчёт по данным MovieLens

## Цель исследования

Цель отчёта — исследовать пользовательские оценки фильмов и определить:

- как распределяются оценки;
- какие пользователи наиболее активны;
- какие фильмы наиболее популярны;
- какие фильмы получили самые высокие и низкие средние оценки;
- какие жанры чаще встречаются среди лучших и худших фильмов;
- как средний рейтинг фильмов меняется в зависимости от десятилетия выпуска.

Для анализа используются данные из файлов `ratings.csv` и `movies.csv`.
Все вычисления выполняются методами модуля `movielens_analysis.py`.

In [1]:
from movielens_analysis import Movies, Ratings

%timeit -n 1 -r 1 from movielens_analysis import Movies, Ratings

875 ns ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


In [2]:
movies = Movies()
ratings = Ratings()

%timeit -n 1 -r 1 Movies(), Ratings()

1.33 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


In [3]:
ratings_list = ratings.get_ratings()
movies_list = movies.get_movies()

%timeit -n 1 -r 1 ratings.get_ratings()
%timeit -n 1 -r 1 movies.get_movies()

32.6 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
6.63 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


## Загрузка данных

Файл `ratings.csv` содержит идентификатор пользователя, идентификатор
фильма и выставленную оценку.

Файл `movies.csv` содержит идентификатор фильма, его название и список
жанров.

На следующем этапе рассмотрим распределение пользовательских оценок.

In [4]:
rating_counts = ratings.count_ratings(ratings_list)

%timeit ratings.count_ratings(ratings_list)

rating_counts

6.76 ms ± 38.3 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Counter({'4.0': 26818,
         '3.0': 20047,
         '5.0': 13211,
         '3.5': 13136,
         '4.5': 8551,
         '2.0': 7551,
         '2.5': 5550,
         '1.0': 2811,
         '1.5': 1791,
         '0.5': 1370})

## Распределение пользовательских оценок

Наиболее часто пользователи ставили оценку 4.0. Реже всего
встречается оценка 0.5.

Распределение показывает, что пользователи чаще ставят
положительные и средние оценки. Это может означать, что ...

In [5]:
active_users = ratings.sort_user(ratings_list)

%timeit ratings.sort_user(ratings_list)

active_users

6.3 ms ± 38.1 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


[('414', 2698),
 ('599', 2478),
 ('474', 2108),
 ('448', 1864),
 ('274', 1346),
 ('610', 1302),
 ('68', 1260),
 ('380', 1218),
 ('606', 1115),
 ('288', 1055)]

## Самые активные пользователи

В результате анализа были определены десять пользователей, которые
оставили больше всего оценок.

Самым активным оказался пользователь с идентификатором 414,
который оценил 2698 фильмов.

Результат показывает, что небольшая группа пользователей может вносить
значительный вклад в формирование базы данных.

In [6]:
sorted_films = movies.get_all_films_information(
    ratings_list,
    movies_list,
)

%timeit -n 1 -r 1 movies.get_all_films_information(ratings_list, movies_list)

sorted_films[:5]

29 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[('Lamerica (1994)',
  {'av_rating': 5.0,
   'reviews': 2,
   'genres': ['Adventure', 'Drama'],
   'year': 1994}),
 ('Heidi Fleiss: Hollywood Madam (1995)',
  {'av_rating': 5.0, 'reviews': 2, 'genres': ['Documentary'], 'year': 1995}),
 ('Awfully Big Adventure, An (1995)',
  {'av_rating': 5.0, 'reviews': 1, 'genres': ['Drama'], 'year': 1995}),
 ('Live Nude Girls (1995)',
  {'av_rating': 5.0, 'reviews': 1, 'genres': ['Comedy'], 'year': 1995}),
 ('In the Realm of the Senses (Ai no corrida) (1976)',
  {'av_rating': 5.0, 'reviews': 1, 'genres': ['Drama'], 'year': 1976})]

## Формирование общей статистики фильмов

Информация из `ratings.csv` и `movies.csv` была объединена по
идентификатору фильма.

Для каждого фильма были рассчитаны:

- средний пользовательский рейтинг;
- количество полученных оценок;
- жанры;
- год выпуска.

Фильмы без пользовательских оценок были исключены из дальнейшего
анализа.

In [7]:
most_reviewed_films = movies.get_most_rev_films(sorted_films)

%timeit -n 1 -r 1 movies.get_most_rev_films(sorted_films)

most_reviewed_films

938 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[('Forrest Gump (1994)',
  {'av_rating': 4.164133738601824,
   'reviews': 329,
   'genres': ['Comedy', 'Drama', 'Romance', 'War'],
   'year': 1994}),
 ('Shawshank Redemption, The (1994)',
  {'av_rating': 4.429022082018927,
   'reviews': 317,
   'genres': ['Crime', 'Drama'],
   'year': 1994}),
 ('Pulp Fiction (1994)',
  {'av_rating': 4.197068403908795,
   'reviews': 307,
   'genres': ['Comedy', 'Crime', 'Drama', 'Thriller'],
   'year': 1994}),
 ('Silence of the Lambs, The (1991)',
  {'av_rating': 4.161290322580645,
   'reviews': 279,
   'genres': ['Crime', 'Horror', 'Thriller'],
   'year': 1991}),
 ('Matrix, The (1999)',
  {'av_rating': 4.192446043165468,
   'reviews': 278,
   'genres': ['Action', 'Sci-Fi', 'Thriller'],
   'year': 1999}),
 ('Star Wars: Episode IV - A New Hope (1977)',
  {'av_rating': 4.231075697211155,
   'reviews': 251,
   'genres': ['Action', 'Adventure', 'Sci-Fi'],
   'year': 1977}),
 ('Jurassic Park (1993)',
  {'av_rating': 3.75,
   'reviews': 238,
   'genres': ['Ac

## Самые популярные фильмы

Популярность фильма определяется количеством пользовательских оценок,
а не значением среднего рейтинга.

Наибольшее количество оценок получил фильм Forrest Gump — его оценили
329 пользователей. Его средний рейтинг составляет 4.16.

Можно заметить, что популярность фильма не обязательно означает
максимальный средний рейтинг.

## Лучшие и худшие фильмы

Фильм с одной оценкой `5.0` нельзя уверенно считать лучшим. Поэтому в
анализ включаются только фильмы, получившие больше 20 пользовательских
оценок.

Лучшие и худшие фильмы определяются по среднему рейтингу.

In [8]:
top_100_films = movies.get_top_100_films(sorted_films)

%timeit -n 1 -r 1 movies.get_top_100_films(sorted_films)

top_100_films[:10]

57.1 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[('Shawshank Redemption, The (1994)',
  {'av_rating': 4.429022082018927,
   'reviews': 317,
   'genres': ['Crime', 'Drama'],
   'year': 1994}),
 ('Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)',
  {'av_rating': 4.333333333333333,
   'reviews': 27,
   'genres': ['Drama', 'Film-Noir', 'Romance'],
   'year': 1950}),
 ('Philadelphia Story, The (1940)',
  {'av_rating': 4.310344827586207,
   'reviews': 29,
   'genres': ['Comedy', 'Drama', 'Romance'],
   'year': 1940}),
 ('In the Name of the Father (1993)',
  {'av_rating': 4.3, 'reviews': 25, 'genres': ['Drama'], 'year': 1993}),
 ('Lawrence of Arabia (1962)',
  {'av_rating': 4.3,
   'reviews': 45,
   'genres': ['Adventure', 'Drama', 'War'],
   'year': 1962}),
 ('Hoop Dreams (1994)',
  {'av_rating': 4.293103448275862,
   'reviews': 29,
   'genres': ['Documentary'],
   'year': 1994}),
 ('Godfather, The (1972)',
  {'av_rating': 4.2890625,
   'reviews': 192,
   'genres': ['Crime', 'Drama'],
   'year': 1972}),
 ('Harold and Maude (1971)',
  {'av_r

### Лучшие фильмы

В список вошли фильмы с наибольшим средним рейтингом, получившие более
20 оценок.

Первое место занимает фильм Побег из Шоушенка со средним рейтингом 4.42.
Количество пользовательских оценок этого фильма — 317.

Использование ограничения по количеству оценок делает результат более
надёжным и исключает фильмы, высокий рейтинг которых основан на
единичных отзывах.

In [9]:
worst_100_films = movies.get_worst_100_films(sorted_films)

%timeit -n 1 -r 1 movies.get_worst_100_films(sorted_films)

worst_100_films[:10]

194 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[('Anaconda (1997)',
  {'av_rating': 1.9259259259259258,
   'reviews': 27,
   'genres': ['Action', 'Adventure', 'Thriller'],
   'year': 1997}),
 ('Godzilla (1998)',
  {'av_rating': 1.9545454545454546,
   'reviews': 33,
   'genres': ['Action', 'Sci-Fi', 'Thriller'],
   'year': 1998}),
 ('Super Mario Bros. (1993)',
  {'av_rating': 2.0,
   'reviews': 21,
   'genres': ['Action',
    'Adventure',
    'Children',
    'Comedy',
    'Fantasy',
    'Sci-Fi'],
   'year': 1993}),
 ('Inspector Gadget (1999)',
  {'av_rating': 2.0952380952380953,
   'reviews': 21,
   'genres': ['Action', 'Adventure', 'Children', 'Comedy'],
   'year': 1999}),
 ('I Know What You Did Last Summer (1997)',
  {'av_rating': 2.109375,
   'reviews': 32,
   'genres': ['Horror', 'Mystery', 'Thriller'],
   'year': 1997}),
 ('Richie Rich (1994)',
  {'av_rating': 2.2,
   'reviews': 25,
   'genres': ['Children', 'Comedy'],
   'year': 1994}),
 ('Wild Wild West (1999)',
  {'av_rating': 2.207547169811321,
   'reviews': 53,
   'genres

### Худшие фильмы

Список сформирован из фильмов с наиболее низким средним рейтингом,
получивших больше 20 пользовательских оценок.

Самый низкий результат у фильма Анаконда. Его средняя оценка составляет
1.92, а количество полученных оценок - 27.

Порог по количеству оценок позволяет считать низкий рейтинг устойчивым,
а не результатом мнения одного пользователя.

In [10]:
interesting_films = movies.get_interesting_films(sorted_films)

%timeit -n 1 -r 1 movies.get_interesting_films(sorted_films)

interesting_films

51.9 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[('Trial, The (Procès, Le) (1962)',
  {'av_rating': 4.9, 'reviews': 5, 'genres': ['Drama'], 'year': 1962}),
 ('Three Billboards Outside Ebbing, Missouri (2017)',
  {'av_rating': 4.75,
   'reviews': 8,
   'genres': ['Crime', 'Drama'],
   'year': 2017}),
 ('Memories of Murder (Salinui chueok) (2003)',
  {'av_rating': 4.7,
   'reviews': 5,
   'genres': ['Crime', 'Drama', 'Mystery', 'Thriller'],
   'year': 2003}),
 ("Swept Away (Travolti da un insolito destino nell'azzurro mare d'Agosto) (1975)",
  {'av_rating': 4.666666666666667,
   'reviews': 6,
   'genres': ['Comedy', 'Drama'],
   'year': 1975}),
 ('Yi Yi (2000)',
  {'av_rating': 4.6, 'reviews': 5, 'genres': ['Drama'], 'year': 2000}),
 ('Secrets & Lies (1996)',
  {'av_rating': 4.590909090909091,
   'reviews': 11,
   'genres': ['Drama'],
   'year': 1996}),
 ('Day of the Doctor, The (2013)',
  {'av_rating': 4.571428571428571,
   'reviews': 7,
   'genres': ['Adventure', 'Drama', 'Sci-Fi'],
   'year': 2013}),
 ("Guess Who's Coming to Dinner

## Малоизвестные высокооценённые фильмы

В этом разделе представлены фильмы, которые получили от 5 до 19 оценок,
но имеют высокий средний рейтинг.

Такие фильмы можно назвать «скрытыми жемчужинами»: они получили
положительные отзывы, но пока не приобрели большой популярности.

Самый высокий средний рейтинг в этой категории у фильма Trial, The (Procès, Le) —
4.9 при 5 пользовательских оценках.

## Анализ жанров

Сначала определим, какие жанры чаще всего встречаются среди ста лучших
и ста худших фильмов.

Если фильм относится к нескольким жанрам, он учитывается один раз в
каждом из них.

In [11]:
top_genres = movies.get_top_genres(top_100_films)

%timeit -n 1 -r 1movies.get_top_genres(top_100_films)

top_genres

[('Drama', 59),
 ('Crime', 29),
 ('Thriller', 29),
 ('Action', 24),
 ('Comedy', 21),
 ('Adventure', 20),
 ('Romance', 19),
 ('War', 17),
 ('Mystery', 13),
 ('Sci-Fi', 11)]

In [12]:
worst_genres = movies.get_worst_genres(worst_100_films)

%timeit -n 1 -r 1 movies.get_worst_genres(worst_100_films)

worst_genres

62.8 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[('Comedy', 51),
 ('Action', 46),
 ('Sci-Fi', 37),
 ('Thriller', 29),
 ('Adventure', 28),
 ('Drama', 17),
 ('Children', 13),
 ('Fantasy', 13),
 ('Horror', 12),
 ('Romance', 12)]

### Жанры лучших и худших фильмов

Наиболее распространённым жанром среди лучших фильмов оказалась
Драма — он встречается в 59 фильмах из ста.

Среди худших фильмов чаще всего встречается жанр Комедия —
он представлен в 51 фильмах.

Некоторые жанры могут одновременно находиться в обоих списках. Это не
обязательно означает, что жанр является хорошим или плохим: такой
результат может быть связан с его общей распространённостью в наборе
данных.

In [13]:
average_genre_ratings = movies.get_av_genres_rating(sorted_films)

%timeit -n 1 -r 1 movies.get_av_genres_rating(sorted_films)

average_genre_ratings.most_common()

3.98 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[('Documentary', 3.781681690126997),
 ('Film-Noir', 3.670470937518959),
 ('War', 3.571654938497546),
 ('Animation', 3.497119150128774),
 ('Drama', 3.4207318655994765),
 ('Western', 3.3833889087601263),
 ('(no genres listed)', 3.3711484593837535),
 ('Romance', 3.364523284189576),
 ('Mystery', 3.330363606538831),
 ('IMAX', 3.3119963710928855),
 ('Crime', 3.3015943279609083),
 ('Musical', 3.2963709618468617),
 ('Fantasy', 3.218843125330384),
 ('Adventure', 3.2152539824965745),
 ('Comedy', 3.1814088447649076),
 ('Thriller', 3.155834205665773),
 ('Children', 3.107690360529312),
 ('Sci-Fi', 3.1024383483769307),
 ('Action', 3.09446807067838),
 ('Horror', 2.918964699710078)]

### Средний рейтинг жанров

Для каждого жанра было рассчитано среднее значение средних рейтингов
всех относящихся к нему фильмов.

Наиболее высокий результат получил жанр Документальное со средним рейтингом
3.78. Самый низкий результат у жанра Хоррор — 2.91.

При интерпретации следует учитывать, что каждый фильм имеет одинаковый
вес независимо от количества пользовательских оценок. Кроме того, один
фильм может одновременно относиться к нескольким жанрам.

## Анализ фильмов по десятилетиям

Фильмы были объединены в группы по десятилетию выпуска. Для каждого
десятилетия рассчитываются:

- среднее значение средних рейтингов фильмов;
- общее количество пользовательских оценок;
- количество оценённых фильмов.

Фильмы, в названии которых отсутствует год выпуска, не учитываются.

In [14]:
year_analytics = movies.get_year_analystics(sorted_films)

%timeit -n 1 -r 1 movies.get_year_analystics(sorted_films)

year_analytics

3.02 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'40-е': {'av_rating': 3.637785645290684, 'reviews': 1101, 'films': 194},
 '50-е': {'av_rating': 3.5754717979099393, 'reviews': 1784, 'films': 277},
 '60-е': {'av_rating': 3.5343531670370707, 'reviews': 2858, 'films': 399},
 '70-е': {'av_rating': 3.4995497664543165, 'reviews': 4994, 'films': 497},
 '80-е': {'av_rating': 3.1699127984361923, 'reviews': 12908, 'films': 1174},
 '90-е': {'av_rating': 3.155819597489887, 'reviews': 37054, 'films': 2205},
 '2000-е': {'av_rating': 3.2598099955186144, 'reviews': 29696, 'films': 2842},
 '2010-е': {'av_rating': 3.2114624435620343, 'reviews': 9482, 'films': 1926}}

### Результаты анализа десятилетий

Наиболее высокий средний рейтинг получили фильмы 40-х годов —
3.63.

Больше всего оценённых фильмов относится к 00-м годам, а наибольшее
количество пользовательских оценок получили фильмы 90-х годов.

Результаты позволяют сравнить не только качество фильмов разных
периодов, но и их представленность и популярность в MovieLens.

Следует учитывать, что средний рейтинг десятилетия рассчитывается как
среднее от средних рейтингов фильмов. Каждый фильм имеет одинаковый вес
независимо от количества пользовательских оценок.

In [15]:
example_year = movies.extract_year("Toy Story (1995)")

%timeit -n 1 -r 1 movies.extract_year("Toy Story (1995)")

example_year

12.3 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


1995

## Извлечение года выпуска

Год выпуска фильма извлекается из его названия. Например, для фильма
`Toy Story (1995)` метод возвращает значение `1995`.

Если название не заканчивается годом в круглых скобках, метод возвращает
`None`.

In [16]:
sorted_ratings = ratings.get_sort_ratings(ratings_list)

%timeit -r 1 -n 1 ratings.get_sort_ratings(ratings_list)

sorted_ratings[:5]

17.3 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[{'user_id': '1', 'movie_id': '47', 'rating': '5.0'},
 {'user_id': '1', 'movie_id': '50', 'rating': '5.0'},
 {'user_id': '1', 'movie_id': '101', 'rating': '5.0'},
 {'user_id': '1', 'movie_id': '151', 'rating': '5.0'},
 {'user_id': '1', 'movie_id': '157', 'rating': '5.0'}]

## Сортировка пользовательских оценок

Пользовательские оценки были отсортированы по убыванию. Первые элементы
списка содержат оценки `5.0`.

Эта сортировка показывает максимальные отдельные оценки, но не
используется для определения лучших фильмов. Для сравнения фильмов
необходимо учитывать их средний рейтинг и количество полученных оценок.

# Итоговые выводы

В ходе исследования были проанализированы пользовательские оценки и
характеристики фильмов из набора данных MovieLens.

Основные результаты:

1. Наиболее распространённой пользовательской оценкой является 4.0.
2. Самым активным оказался пользователь 414, оставивший 2698
   оценок.
3. Самым популярным по количеству оценок стал фильм Форрест Гамп.
4. Наиболее высокий устойчивый средний рейтинг среди фильмов с более
   чем 20 оценками получил фильм Побег из Шоушенка.
5. Наиболее низкий средний рейтинг среди фильмов с таким же ограничением
   получил фильм Анаконда.
6. Среди лучших фильмов чаще всего встречается жанр Драма.
7. Самый высокий средний рейтинг среди жанров получил Документальное.
8. Наиболее высокий средний рейтинг характерен для фильмов 40-х
   годов.

Исследование показывает, что популярность и средний рейтинг — разные
характеристики. Фильм может получить большое количество оценок, но не
занимать первое место по их среднему значению. Ограничение по количеству
оценок помогает получать более устойчивые результаты и уменьшает
влияние случайных единичных отзывов.